chain-impl-mockchain/src/header/cstruct.rs

```diff
@@ -0,0 +1,419 @@
+    pub fn set_parent_hash(&mut self, s: &ParentHash) {
+        unsafe {
+            self.unsigned[HEADER_OFFSET_PARENT_HASH..HEADER_COMMON_SIZE].copy_from_slice(&s[..])
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn set_bft_leader_id(&mut self, s: &BftLeaderId) {
+        assert_eq!(self.version(), VERSION_BFT);
+        unsafe {
+            self.bft[HEADER_OFFSET_BFT_LEADER_ID..HEADER_OFFSET_BFT_SIGNATURE]
+                .copy_from_slice(&s[..])
+        }
+    }
+
+    pub fn set_bft_leader_id_slice(&mut self, s: &[u8]) {
+        assert_eq!(self.version(), VERSION_BFT);
+        assert_eq!(s.len(), size_of::<BftLeaderId>());
+        unsafe {
+            self.bft[HEADER_OFFSET_BFT_LEADER_ID..HEADER_OFFSET_BFT_SIGNATURE].copy_from_slice(s)
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn set_bft_signature(&mut self, s: &BftSignature) {
+        assert_eq!(self.version(), VERSION_BFT);
+        unsafe { self.bft[HEADER_OFFSET_BFT_SIGNATURE..HEADER_BFT_SIZE].copy_from_slice(&s[..]) }
+    }
+
+    pub fn set_bft_signature_slice(&mut self, s: &[u8]) {
+        assert_eq!(self.version(), VERSION_BFT);
+        assert_eq!(s.len(), size_of::<BftSignature>());
+        unsafe { self.bft[HEADER_OFFSET_BFT_SIGNATURE..HEADER_BFT_SIZE].copy_from_slice(s) }
+    }
+
+    pub fn set_gp_node_id(&mut self, s: &GpNodeId) {
+        assert_eq!(self.version(), VERSION_GP);
+        unsafe { self.gp[HEADER_OFFSET_GP_ID..HEADER_OFFSET_GP_VRF_PROOF].copy_from_slice(&s[..]) }
+    }
+
+    #[allow(dead_code)]
+    pub fn set_gp_node_id_slice(&mut self, s: &[u8]) {
+        assert_eq!(self.version(), VERSION_GP);
+        assert_eq!(s.len(), size_of::<GpNodeId>());
+        unsafe { self.gp[HEADER_OFFSET_GP_ID..HEADER_OFFSET_GP_VRF_PROOF].copy_from_slice(s) }
+    }
+
+    pub fn set_gp_vrf_proof(&mut self, s: &GpVrfProof) {
+        assert_eq!(self.version(), VERSION_GP);
+        unsafe {
+            self.gp[HEADER_OFFSET_GP_VRF_PROOF..HEADER_OFFSET_GP_KES_SIG].copy_from_slice(&s[..])
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn set_gp_vrf_proof_slice(&mut self, s: &[u8]) {
+        assert_eq!(self.version(), VERSION_GP);
+        assert_eq!(s.len(), size_of::<GpVrfProof>());
+        unsafe { self.gp[HEADER_OFFSET_GP_VRF_PROOF..HEADER_OFFSET_GP_KES_SIG].copy_from_slice(s) }
+    }
+
+    #[allow(dead_code)]
+    pub fn set_gp_kes_signature(&mut self, s: &GpKesSignature) {
+        assert_eq!(self.version(), VERSION_GP);
+        unsafe { self.gp[HEADER_OFFSET_GP_KES_SIG..HEADER_GP_SIZE].copy_from_slice(&s[..]) }
+    }
+
+    pub fn set_gp_kes_signature_slice(&mut self, s: &[u8]) {
+        assert_eq!(self.version(), VERSION_GP);
+        assert_eq!(s.len(), size_of::<GpKesSignature>());
+        unsafe { self.gp[HEADER_OFFSET_GP_KES_SIG..HEADER_GP_SIZE].copy_from_slice(s) }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum HeaderError {
+    InvalidSize,
+    UnknownVersion,
+    SizeMismatch { expected: usize, got: usize },
+}
```

Comment:
missing `Display` and `Error` implementation. This type is not going to propagate well to the user of the APIs, creating.